source-mongodb/oplog.go

```diff
@@ -0,0 +1,44 @@
+package main
+
+import (
+	"context"
+	"fmt"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+type oplogRecord struct {
+	Ts primitive.Timestamp `bson:"ts"`
+}
+
+const tsProperty = "ts"
+
+// Oplog time difference: the time difference is measured as the
+// difference between the latest and the oldest record in a MongoDB oplog, and
+// it is used as a measure of how long do oplog records stay in the oplog.
+// Note that this value can fluctuate depending on the density of oplog
+// records during a period, so continuously monitoring this value is necessary
+// and a single value is not representative
+func OplogTimeDifference(ctx context.Context, client *mongo.Client) (uint32, error) {
+	var db = client.Database("local")
+	var oplog = db.Collection("oplog.rs")
+
+	// get latest record
+	var opts = options.FindOne().SetSort(bson.D{{tsProperty, SortDescending}})
```

Comment:
A single record would just return `0` as the time difference, and no records would error.

I think having no records is _theoretically_ possible. How could we handle it? We could log a warn that we could not read the oplog time difference if we can't read it, instead of erroring.

Reply:
Yeah I think we'd have to log a warning instead of erroring in that case. I could imagine a static database that had been populated longer than `{oplog_retention_period}` ago with no changes since then, and we would want to be able to capture from such a database.